crates/sp-domains-fraud-proof/src/runtime_interface.rs

```diff
@@ -75,12 +76,14 @@ pub trait FraudProofRuntimeInterface {
     fn construct_domain_inherent_extrinsic(
         &mut self,
         domain_runtime_code: Vec<u8>,
+        invalid_inherent_extrinsic_data: InvalidInherentExtrinsicData,
         domain_inherent_extrinsic_data: DomainInherentExtrinsicData,
     ) -> Option<DomainInherentExtrinsic> {
         self.extension::<FraudProofExtension>()
             .expect("No `FraudProofExtension` associated for the current context!")
             .construct_domain_inherent_extrinsic(
                 domain_runtime_code,
+                invalid_inherent_extrinsic_data,
```

Comment:
We need to be extremely careful with the host function change as mainnet launch because the host function signature is compiled into the runtime code and the runtime will crash if the exact same host function does not exist in the client side, and for the node that sync from genesis it will go through all version of runtime code and expect all the host function exist. 

To properly handle host function change we need to:
- Keep the old version of host function signature untouched
    - If the host function is ever used within the blockcain history, we can change the content of the host function but must ensure it will return the exact same value as before
    - If the host function is never used within the blockcain history, we can safely replace the content with dummy return value
- Introduce the new host function with `register_only` attribute, and versioning if it has the same name as other existing host function, e.g. `#[version(2, register_only)]` also see #2755
- Wait until most of the node in the network upgraded to release the contain the new host function in the client side
- Remove `register_only` and use the new host function in the runtime, upgrade runtime code

Reply:
While Ning is correct in terms of versioning the Host functions when the signature changes.
in this scenario, @NingLin-P are we using this host function any where? Domains are not launched yet so I believe no one is using it yet.

as long as this changes goes into the network before the domains are initialized, this should not be an issue. Correct me if I'm missing something here

Reply:
>as long as this changes goes into the network before the domains are initialized, this should not be an issue.

The host function is used in the consensus runtime (not domain runtime) for FP verification and is already part of the consensus runtime (included by `pallet-domains`) in the mainnet, so it is already a compatibility issue we need to consider.

We are only safe to change the content of the host function as there is no domain initialized and the host function has never been called yet.

Reply:
Ahh yes. Correct ! 💯

Reply:
There’s no need to change the signature of this particular function. Instead, I can:
* revert the changes to `DomainInherentExtrinsicData`
* return a new struct with fewer fields from `InvalidInherentExtrinsicProof::verify()`
* copy the necessary fields into `DomainInherentExtrinsicData` before calling runtime code

This change will also need to be made in PRs #3315 and #3320.

In this PR, can I change the name of a field in a struct that is passed to runtime code, or should I leave the name exactly the same?
(We replaced the block randomness with the extrinsics shuffling seed, but the type is the same, and it isn’t actually used by the runtime code.)

Reply:
>In this PR, can I change the name of a field in a struct that is passed to runtime code, or should I leave the name exactly the same?
(We replaced the block randomness with the extrinsics shuffling seed, but the type is the same, and it isn’t actually used by the runtime code.)

TBH, I'm not 100% sure (perhaps @nazar-pc knows), better to verify before any changes as it can be fatal to the network, also see [the upstream doc](https://github.com/paritytech/polkadot-sdk/blob/master/substrate/primitives/runtime-interface/src/lib.rs).

Reply:
Structs are sent to the runtime in SCALE encoding and the name of the field doesn't matter there.

Reply:
Just in case, I created a `host-function-change` label, PR attached with this label must test compatibility with the existing network before release. cc @dariolina @jfrank-summit

Reply:
Changing the name of a struct field is ok as long as the type of the field and struct itself does not change.

I would recommend create a new version of host function so that we do not have to patch this existing one